Source/bmalloc/libpas/src/libpas/pas_scavenger.c

```diff
@@ -387,6 +400,35 @@ static void* scavenger_thread_main(void* arg)
     return NULL;
 }
 
+bool pas_scavenger_try_install_foreign_work_callback(
+    pas_scavenger_foreign_work_callback callback,
+    uint32_t period_log2_ms,
+    void* userdata)
+{
+    pas_scavenger_data* data;
+
+    PAS_ASSERT(callback);
+
+    data = ensure_data_instance(pas_lock_is_not_held);
+    pthread_mutex_lock(&data->foreign_work.lock);
+
+    int slot = data->foreign_work.next_open_descriptor;
+    if (slot >= PAS_SCAVENGER_MAX_FOREIGN_WORK_DESCRIPTORS)
+        return false;
+
+    double requested_period_ms = pow(2.0, period_log2_ms);
+    uint32_t requested_ticks = (uint32_t)(requested_period_ms / pas_scavenger_period_in_milliseconds);
+
+    data->foreign_work.descriptors[slot].period_log2_ticks = pas_log2(requested_ticks);
+    data->foreign_work.descriptors[slot].func = callback;
+    data->foreign_work.descriptors[slot].userdata = userdata;
```

Comment:
Emit store-barrier here. So scavenger can see the fully stored slot when `next_open_descriptor` includes this slot.